import_export/widgets.py

```diff
@@ -371,10 +371,12 @@ class Meta:
 
     :param model: The Model the ForeignKey refers to (required).
     :param field: A field on the related model used for looking up a particular object.
+    :param use_natural_foreign_keys: Boolean for whether to use Django natural key functions, default False
     """
-    def __init__(self, model, field='pk', *args, **kwargs):
+    def __init__(self, model, field='pk', use_natural_foreign_keys=False, *args, **kwargs):
```

Comment:
My suggestion is to move this to a separate `ForeignKeyWidget` subclass to avoid impacting existing users (see [comment](https://github.com/django-import-export/django-import-export/pull/1371#issuecomment-1001748740))

Reply:
So my reasoning for doing this (and not just making my own widget that inherits from foreignkeywidget) is that the way widgets are structured you have to duplicate almost the entire logic and that's a pretty horrible DRY violation -- I found that when I was writing my own subclassed widgets and fields that subclassing them meant usually rewriting the entire clean and render method (or save in the case of fields), which would then be out of date when the parents were updated.  

In Render specifically you have the whole attrs splitting logic and the try/except block that then has to be changed multiple places if Django changes.

The other major advantage to using ForeignKeyWidget is that if we were to add a  _meta flag to the resourceoptions, we could force an entire model to default to using natural keys without having to write another widget detection function. I think that'd make for quite a bit cleaner interface with less to maintain than another widget detection function, another widget, and so on.  

I do not think this could break any existing subclassed ForeignKeyWidgets unless someone explicity passed it a parameter it couldn't support, but might be missing something. 

I think the functionality of using django natural key functions is useful enough that I'm happy to rewrite it that way I just worry about how wet that widget will be. It wouldn't be able to use the inherited widget's functionality outside of the __init__ (whereas with this design it reuses most of the code).

Reply:
OK thanks - I understand your point of view here.  Although  I don't understand your proposal re the _meta flag, so you'd have to clarify that for me.  My only concern is the risk of breaking existing implementations, but I think as long as users don't pass the optional param (as you say) then it should be ok.  Perhaps you could help me understand the additional use of _meta?

Reply:
Sure! 
Suppose we add a meta option to our resourceoptions class that mirrors the Django flag 'use_natural_foreing_keys' so that a ModelResource can set in its Meta "use_natural_foreign_keys." Then we can update the get_fk_widget to check that option, and pass use_natural_foreign_keys = True to all detected foreign key widgets *if the models have natural key functions*. 

It feels more "natural" to me to allow fitting it into the auto detection of widgets logic in the future. get_fk_widget could say (in shorthand), "if my _meta option use_natural_foreign_keys is True, and the field class has a natural key function, and the field has a manager with a get by natural key function, then pass use_natural_foreign_keys to that foreign key widget automatically. 
 
Then you could tell a model resource to use natural keys wherever it can without needing to declare every single field, it'd just try to use them wherever it could. 

(This could be accomplished with adding an additional widget, just requires functionally duplicating the get_fk_widget function vs. extending it).

Reply:
Thanks Ryan - leave it with me and I'll take a look.  Am pretty busy for the next few days but I'll look into it when I can.

Reply:
Thanks Matthew. I'm more than happy to do another PR for the resource option update to allow global use of natural keys on a resource! 

The thing I am stuck on the most is figuring out how I would move the model managers into the tests area, I don't see any clean way to do that. 

I could obscure it a little by making a managers.py in the reference app and importing them possibly, but models seem to have to have the manager reference defined in their code.  

Let me know if there's anything else I can dig into.

Reply:
Hi Ryan - would you mind clarifying the use case for these changes?  Presumably it is required when you are importing related models defined in JSON files?
thanks

Reply:
Sure!

The major use case for using natural keys is moving data between Django environments -- so that if your numbering is different in one environment you can still load data with relationships by natural keys. In the book example, if GRRM is id 1 in environment TEST, and id 2 in environment PROD, you can still load a book with a natural key relationship. 

Using natural keys instead of single fields (e.g. name, like the author example) you could use more complex keys. In real life, an author likely is not unique by name, but by Name + some other identifying characteristic. 

The Book natural key example of author+name shows that a little more; so if you had say, an Edition that relates to Book it could relate to the book by its author+name instead of just by Name (since there are, realistically, a lot of books with duplicate titles). 

This is particularly useful if you have environmens with subsets of data - say you have a test env with only 100 books, and a production environment with millions of books, the numbering would be different. But if you wanted to bring a few new books from Prod into your test environment, you could do that even if the authors they relate to already exist with different PK IDs than in prod. 

TL;DR Natural keys are most useful for dealing with moving data between environments where the PK numbers are not synchronized and especially when natural keys are more than a single field, so you can't reliably identify relations by PK id.

Reply:
thanks for the explanation.  It seems like a useful addition.  Having a meta attribute which enables it as you describe would be good.  However, the only caveats are that I would like this to have no impact on the existing example application, and it will need to be documented appropriately, perhaps as an addition to the existing rst docs.  Is there anything blocking you from proceeding?

Reply:
I can't find any clear cut way to put the managers/model changes into the tests -- the models don't get created, and I can't attach a modelmanager class to an existing model. If you can help me figure out how to do that I can move the models and managers, but I tried for a while and could not think of a way to do it in the tests. 

There is zero functional impact to the example application having the natural key functions. I could make the Manager classes hidden by moving them into a managers.py and then importing them. 

Is there something I need to do beyond updating the docstring of the foreignkeywidget, documentationwise? I made some updates to that that should have more info on how to use it. I could make similar updates but I assumed they were autogenerated from the docstrings.

Reply:
Thanks for looking into it.  I think we will have to have the ModelManagers added to the example application as you have it in the PR.  It's not ideal but perhaps you could add a comment to the BookManager docstring clarifying that it is only necessary if natural key functionality is required.

It would be great if you could add a short section to the rst docs indicating how natural key functionality can be implemented.  Most users seeking this functionality will read the docs rather than the code.

Reply:
Sure, I will add some comments to that effect on all the natural key sections. 

If you know which RST you think the natural key stuff should be in I'm happy to do it.